File: Deepseek-Coder-33b-Instruct/stage2/data_deviceptr_clause.cpp
#include <stdlib.h>
#include <stdio.h>
#include <openacc.h>

#define N 100

int main() {
    int a[N], a_device[N];
    int i, err = 0;

    // Initialize the array on the host
    for (i = 0; i < N; i++) {
        a[i] = i;
    }

    // Copy the array to the device
    #pragma acc data copyin(a[:N]) copyout(a_device[:N])
    {
        // Modify the array on the device
        #pragma acc kernels
        for (i = 0; i < N; i++) {
            a_device[i] = a[i] * 2;
        }
    }

    // Compare the modified array with the original array
    for (i = 0; i < N; i++) {
        if (a_device[i] != a[i] * 2) {
            err++;
        }
    }

    // Print the result
    if (err == 0) {
        printf("Test passed.\n");
    } else {
        printf("Test failed.\n");
    }

    return err;
}

File: Deepseek-Coder-33b-Instruct/stage2/data_deviceptr_clause.f90
!T1:data deviceptr clause,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER :: errors = 0
        INTEGER, PARAMETER :: N = 100
        INTEGER :: i
        INTEGER, DIMENSION(N) :: host_data, device_data

        ! Initialize host data
        DO i = 1, N
          host_data(i) = i
        END DO

        ! Copy data to device
        !$acc data copyin(host_data) copyout(device_data)
        !$acc update device(host_data)

        ! Use device data
        !$acc parallel loop deviceptr(host_data)
        DO i = 1, N
          device_data(i) = host_data(i)
        END DO

        !$acc end data

        ! Check results
        DO i = 1, N
          IF (host_data(i) .ne. device_data(i)) THEN
            errors = errors + 1
          END IF
        END DO

        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
        LOGICAL :: test1
        failed = .FALSE.
        failcode = 0
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
        CALL EXIT (failcode)
      END PROGRAM